### PR TITLE
H264 encoding

### PR DESCRIPTION
**Pull request by valegagge**
Here is implemented the H264 carrier. 
It can be used to read and decode a h264 stream published by Gstreamer.

In order to accomplished it, I needed to add method `yarp::os::carrier::createFace()`,  that returns the needed face of the carrier. This method is used in `yarp::os::carriers::listen` and in `yarp::os::carriers::connect `  in order to open new connection using the correct face.

### CODE REVIEW

**Comment by drdanz** [comment]
Sorry @valegagge for the HUGE delay...

* Rebased
* Fixed CMake stuff
* Small cleanup and fix copyright headers
* Try to enable the carrier on travis/linux

Besides this, it looks good to me, just some comments for future reference, it would be nice to have these informations somewhere in the documentation:

* Packages required to **build** on debian testing
  * `libgstreamer1.0-dev`
  * `libgstreamer-plugins-base1.0-dev`
* Packages required to **run** on debian testing (untested)
  * `gstreamer1.0-plugins-base` (for `videoconvert` and `appsink`)
  * `gstreamer1.0-plugins-good` (for `udpsrc` and `rtph264depay`)
  * `gstreamer1.0-plugins-bad` (for `h264parse`)
  * `gstreamer1.0-libav` (for `avdec_h264`)
* Useful packages but not required on debian testing
  * `gstreamer1.0-tools`

**Comment by drdanz** [comment]
The error on MacOS is no longer related to CMake:

```
[ 46%] Linking CXX shared module ../../../lib/yarp/yarp_human.so
Undefined symbols for architecture x86_64:
  "ACE_SOCK_Acceptor::ACE_SOCK_Acceptor()", referenced from:
      yarp::os::impl::TcpFace::TcpFace() in HumanCarrier.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/yarp/yarp_human.so] Error 1
make[1]: *** [src/carriers/human_carrier/CMakeFiles/yarp_human.dir/all] Error 2
make: *** [all] Error 2
```

@valegagge Perhaps you can ask @francesco-romano or someone else with a mac?

**Comment by traversaro** [comment]
Test on macOS : the problem is that by including `#include <yarp/os/impl/TcpFace.h>` in `Carriers.h` we are including ACE headers in public YARP headers (depending on `YARP_HAS_ACE`), see https://github.com/robotology/yarp/pull/1346/files#diff-860774c077ca7454a1b62140bf0cef3bR19 .

**Comment by valegagge** [comment]
Thanks @traversaro .

The solution could be move the implementation of functions in a cpp file. Moreover this is a good practice in code writing.

File: src/libYARP_OS/include/yarp/os/Carrier.h [hunk 1]
```diff
@@ -15,6 +15,8 @@
 #include <yarp/os/ConnectionReader.h>
 #include <yarp/os/Connection.h>
 #include <yarp/os/ConnectionState.h>
+#include <yarp/os/Face.h>
+#include <yarp/os/impl/TcpFace.h> // to move in cpp file
```
Comment (drdanz) [inline src/libYARP_OS/include/yarp/os/Carrier.h]: Sorry, I just noticed this after @traversaro comment, it is not ok to have "impl" headers included in "not-impl" headers, please move this in the .cpp file

File: src/libYARP_OS/include/yarp/os/Carrier.h [hunk 2]
```diff
@@ -520,6 +522,14 @@ class YARP_OS_API yarp::os::Carrier : public Connection
         YARP_UNUSED(params);
     }
 
+    /**
+     * Create new Face object that the carrier needs.
+     *
+     */
+    virtual yarp::os::Face* createFace(void)
```
Comment (drdanz) [inline src/libYARP_OS/include/yarp/os/Carrier.h]: I wonder if this method would make sense in yarp::os::Connection

File: src/libYARP_OS/include/yarp/os/Carrier.h (hunk 1 above)
Comment (valegagge) [reply src/libYARP_OS/include/yarp/os/Carrier.h]: All functions have been implemented in header file. So I'll move all implementations to .cpp file.

File: src/libYARP_OS/include/yarp/os/Carrier.h (hunk 2 above)
Comment (valegagge) [reply src/libYARP_OS/include/yarp/os/Carrier.h]: I think that the connection is the "pipe" between two ports,  while the carrier specifies the type of data that passes through the pipe and the face is the network support on wich data relies on.
So the face dependes on carrier, and I think it is better mantain the face in the carrier.

**Comment by valegagge** [comment]
Hi @drdanz,

I try to compile this branch, but I obtain following error 
` 87%] Building CXX object src/carriers/h264_carrier/CMakeFiles/yarp_h264.dir/H264Decoder.cpp.o
In file included from /usr/include/gstreamer-1.0/gst/gstbin.h:27:0,
                 from /usr/include/gstreamer-1.0/gst/gst.h:35,
                 from /usr/local/src/robot/yarp/src/carriers/h264_carrier/H264Decoder.cpp:12:
/usr/include/gstreamer-1.0/gst/gstelement.h:55:27: fatal error: gst/gstconfig.h: No such file or directory
compilation terminated.
src/carriers/h264_carrier/CMakeFiles/yarp_h264.dir/build.make:134: recipe for target 'src/carriers/h264_carrier/CMakeFiles/yarp_h264.dir/H264Decoder.cpp.o' failed
make[2]: *** [src/carriers/h264_carrier/CMakeFiles/yarp_h264.dir/H264Decoder.cpp.o] Error 1
CMakeFiles/Makefile2:2360: recipe for target 'src/carriers/h264_carrier/CMakeFiles/yarp_h264.dir/all' failed
make[1]: *** [src/carriers/h264_carrier/CMakeFiles/yarp_h264.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
`

I verified that the file required 'gst/gstconfig.h' is in /usr/lib/x86_64-linux-gnu/gstreamer-1.0/include/gst/gstconfig.h, so maybe we need to add plugins' include directory also.

The strange thing is that this error is not found by travis even if the configuration is '-DENABLE_yarpcar_h264:BOOL=ON'

File: src/libYARP_OS/include/yarp/os/Carrier.h (hunk 1 above)
Comment (valegagge) [reply src/libYARP_OS/include/yarp/os/Carrier.h]: Done. see 7d355df16ee4f1f68a34482e96175836f1dceec9

**Comment by drdanz** [comment]
@valegagge That's weird, I don't have this error, and it does not happen on travis...
Can you please check in the cmake output if you have something like this?
```
--  +++ GStreamer (>= 1.4): found (/usr/lib/x86_64-linux-gnu/libgstreamer-1.0.so)
--  +++ GStreamerPluginsBase (>= 1.4): found
```

Also can you post the output of `cat CMakeCache.txt | grep -i GST`?

Finally can you run `VERBOSE=1 make` and post the command line executed when your build fails?

**Comment by valegagge** [comment]
I have 
```
--  +++ GStreamer (>= 1.4): found (/usr/lib/x86_64-linux-gnu/libgstreamer-1.0.so)
--  +++ GStreamerPluginsBase (>= 1.4): found
```

 `cat CMakeCache.txt | grep -i GST` :

```
GSTREAMER_APP_INCLUDE_DIR:PATH=/usr/include/gstreamer-1.0
GSTREAMER_APP_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libgstapp-1.0.so
GSTREAMER_INCLUDE_DIR:PATH=/usr/include/gstreamer-1.0
GSTREAMER_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libgstreamer-1.0.so
//The path to the gstreamer plugins installation directory
GSTREAMER_PLUGIN_DIR:PATH=/usr/lib/x86_64-linux-gnu/gstreamer-1.0
//Use package GStreamer
YARP_USE_GSTREAMER:BOOL=ON
//Use package GStreamerPluginsBase
YARP_USE_GSTREAMERPLUGINSBASE:BOOL=ON
yarp_h264_LIB_DEPENDS:STATIC=general;YARP::YARP_OS;general;YARP::YARP_sig;general;YARP::YARP_wire_rep_utils;general;/usr/lib/x86_64-linux-gnu/libglib-2.0.so;general;/usr/lib/x86_64-linux-gnu/libgstreamer-1.0.so;
//Details about finding GStreamer
FIND_PACKAGE_MESSAGE_DETAILS_GStreamer:INTERNAL=[/usr/lib/x86_64-linux-gnu/libgstreamer-1.0.so][/usr/include/gstreamer-1.0][TRUE][v(1.4)]
//ADVANCED property for variable: GSTREAMER_APP_INCLUDE_DIR
GSTREAMER_APP_INCLUDE_DIR-ADVANCED:INTERNAL=1
//ADVANCED property for variable: GSTREAMER_APP_LIBRARY
GSTREAMER_APP_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: GSTREAMER_INCLUDE_DIR
GSTREAMER_INCLUDE_DIR-ADVANCED:INTERNAL=1
//ADVANCED property for variable: GSTREAMER_LIBRARY
GSTREAMER_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: GSTREAMER_PLUGIN_DIR
GSTREAMER_PLUGIN_DIR-ADVANCED:INTERNAL=1
PKG_GSTREAMER_APP_CFLAGS:INTERNAL=-pthread;-I/usr/include/gstreamer-1.0;-I/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;-I/usr/include/glib-2.0;-I/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_APP_CFLAGS_I:INTERNAL=
PKG_GSTREAMER_APP_CFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_APP_FOUND:INTERNAL=1
PKG_GSTREAMER_APP_INCLUDEDIR:INTERNAL=/usr/include/gstreamer-1.0
PKG_GSTREAMER_APP_INCLUDE_DIRS:INTERNAL=/usr/include/gstreamer-1.0;/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;/usr/include/glib-2.0;/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_APP_LDFLAGS:INTERNAL=-lgstapp-1.0;-lgstbase-1.0;-lgstreamer-1.0;-lgobject-2.0;-lglib-2.0
PKG_GSTREAMER_APP_LDFLAGS_OTHER:INTERNAL=
PKG_GSTREAMER_APP_LIBDIR:INTERNAL=/usr/lib/x86_64-linux-gnu
PKG_GSTREAMER_APP_LIBRARIES:INTERNAL=gstapp-1.0;gstbase-1.0;gstreamer-1.0;gobject-2.0;glib-2.0
PKG_GSTREAMER_APP_LIBRARY_DIRS:INTERNAL=
PKG_GSTREAMER_APP_LIBS:INTERNAL=
PKG_GSTREAMER_APP_LIBS_L:INTERNAL=
PKG_GSTREAMER_APP_LIBS_OTHER:INTERNAL=
PKG_GSTREAMER_APP_LIBS_PATHS:INTERNAL=
PKG_GSTREAMER_APP_PREFIX:INTERNAL=/usr
PKG_GSTREAMER_APP_STATIC_CFLAGS:INTERNAL=-pthread;-I/usr/include/gstreamer-1.0;-I/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;-I/usr/include/glib-2.0;-I/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_APP_STATIC_CFLAGS_I:INTERNAL=
PKG_GSTREAMER_APP_STATIC_CFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_APP_STATIC_INCLUDE_DIRS:INTERNAL=/usr/include/gstreamer-1.0;/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;/usr/include/glib-2.0;/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_APP_STATIC_LDFLAGS:INTERNAL=-lgstapp-1.0;-lgstbase-1.0;-lgstreamer-1.0;-lgmodule-2.0;-pthread;-ldl;-lgobject-2.0;-lffi;-lglib-2.0;-pthread;-lpcre;-pthread
PKG_GSTREAMER_APP_STATIC_LDFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_APP_STATIC_LIBDIR:INTERNAL=
PKG_GSTREAMER_APP_STATIC_LIBRARIES:INTERNAL=gstapp-1.0;gstbase-1.0;gstreamer-1.0;gmodule-2.0;dl;gobject-2.0;ffi;glib-2.0;pcre
PKG_GSTREAMER_APP_STATIC_LIBRARY_DIRS:INTERNAL=
PKG_GSTREAMER_APP_STATIC_LIBS:INTERNAL=
PKG_GSTREAMER_APP_STATIC_LIBS_L:INTERNAL=
PKG_GSTREAMER_APP_STATIC_LIBS_OTHER:INTERNAL=
PKG_GSTREAMER_APP_STATIC_LIBS_PATHS:INTERNAL=
PKG_GSTREAMER_APP_VERSION:INTERNAL=1.8.3
PKG_GSTREAMER_APP_gstreamer-app-1.0_INCLUDEDIR:INTERNAL=
PKG_GSTREAMER_APP_gstreamer-app-1.0_LIBDIR:INTERNAL=
PKG_GSTREAMER_APP_gstreamer-app-1.0_PREFIX:INTERNAL=
PKG_GSTREAMER_APP_gstreamer-app-1.0_VERSION:INTERNAL=
PKG_GSTREAMER_CFLAGS:INTERNAL=-pthread;-I/usr/include/gstreamer-1.0;-I/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;-I/usr/include/glib-2.0;-I/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_CFLAGS_I:INTERNAL=
PKG_GSTREAMER_CFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_FOUND:INTERNAL=1
PKG_GSTREAMER_INCLUDEDIR:INTERNAL=/usr/include/gstreamer-1.0
PKG_GSTREAMER_INCLUDE_DIRS:INTERNAL=/usr/include/gstreamer-1.0;/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;/usr/include/glib-2.0;/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_LDFLAGS:INTERNAL=-lgstreamer-1.0;-lgobject-2.0;-lglib-2.0
PKG_GSTREAMER_LDFLAGS_OTHER:INTERNAL=
PKG_GSTREAMER_LIBDIR:INTERNAL=/usr/lib/x86_64-linux-gnu
PKG_GSTREAMER_LIBRARIES:INTERNAL=gstreamer-1.0;gobject-2.0;glib-2.0
PKG_GSTREAMER_LIBRARY_DIRS:INTERNAL=
PKG_GSTREAMER_LIBS:INTERNAL=
PKG_GSTREAMER_LIBS_L:INTERNAL=
PKG_GSTREAMER_LIBS_OTHER:INTERNAL=
PKG_GSTREAMER_LIBS_PATHS:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_CFLAGS:INTERNAL=-pthread;-I/usr/include/gstreamer-1.0;-I/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;-I/usr/include/glib-2.0;-I/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_PLUGINS_BASE_CFLAGS_I:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_CFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_PLUGINS_BASE_FOUND:INTERNAL=1
PKG_GSTREAMER_PLUGINS_BASE_INCLUDEDIR:INTERNAL=/usr/include/gstreamer-1.0
PKG_GSTREAMER_PLUGINS_BASE_INCLUDE_DIRS:INTERNAL=/usr/include/gstreamer-1.0;/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;/usr/include/glib-2.0;/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_PLUGINS_BASE_LDFLAGS:INTERNAL=-lgstreamer-1.0;-lgobject-2.0;-lglib-2.0
PKG_GSTREAMER_PLUGINS_BASE_LDFLAGS_OTHER:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_LIBDIR:INTERNAL=/usr/lib/x86_64-linux-gnu
PKG_GSTREAMER_PLUGINS_BASE_LIBRARIES:INTERNAL=gstreamer-1.0;gobject-2.0;glib-2.0
PKG_GSTREAMER_PLUGINS_BASE_LIBRARY_DIRS:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_LIBS:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_LIBS_L:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_LIBS_OTHER:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_LIBS_PATHS:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_PREFIX:INTERNAL=/usr
PKG_GSTREAMER_PLUGINS_BASE_STATIC_CFLAGS:INTERNAL=-pthread;-I/usr/include/gstreamer-1.0;-I/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;-I/usr/include/glib-2.0;-I/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_PLUGINS_BASE_STATIC_CFLAGS_I:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_STATIC_CFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_PLUGINS_BASE_STATIC_INCLUDE_DIRS:INTERNAL=/usr/include/gstreamer-1.0;/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;/usr/include/glib-2.0;/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LDFLAGS:INTERNAL=-lgstreamer-1.0;-lgmodule-2.0;-pthread;-ldl;-lgobject-2.0;-lffi;-lglib-2.0;-pthread;-lpcre;-pthread
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LDFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LIBDIR:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LIBRARIES:INTERNAL=gstreamer-1.0;gmodule-2.0;dl;gobject-2.0;ffi;glib-2.0;pcre
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LIBRARY_DIRS:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LIBS:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LIBS_L:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LIBS_OTHER:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_STATIC_LIBS_PATHS:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_VERSION:INTERNAL=1.8.3
PKG_GSTREAMER_PLUGINS_BASE_gstreamer-plugins-base-1.0_INCLUDEDIR:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_gstreamer-plugins-base-1.0_LIBDIR:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_gstreamer-plugins-base-1.0_PREFIX:INTERNAL=
PKG_GSTREAMER_PLUGINS_BASE_gstreamer-plugins-base-1.0_VERSION:INTERNAL=
PKG_GSTREAMER_PREFIX:INTERNAL=/usr
PKG_GSTREAMER_STATIC_CFLAGS:INTERNAL=-pthread;-I/usr/include/gstreamer-1.0;-I/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;-I/usr/include/glib-2.0;-I/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_STATIC_CFLAGS_I:INTERNAL=
PKG_GSTREAMER_STATIC_CFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_STATIC_INCLUDE_DIRS:INTERNAL=/usr/include/gstreamer-1.0;/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include;/usr/include/glib-2.0;/usr/lib/x86_64-linux-gnu/glib-2.0/include
PKG_GSTREAMER_STATIC_LDFLAGS:INTERNAL=-lgstreamer-1.0;-lgmodule-2.0;-pthread;-ldl;-lgobject-2.0;-lffi;-lglib-2.0;-pthread;-lpcre;-pthread
PKG_GSTREAMER_STATIC_LDFLAGS_OTHER:INTERNAL=-pthread
PKG_GSTREAMER_STATIC_LIBDIR:INTERNAL=
PKG_GSTREAMER_STATIC_LIBRARIES:INTERNAL=gstreamer-1.0;gmodule-2.0;dl;gobject-2.0;ffi;glib-2.0;pcre
PKG_GSTREAMER_STATIC_LIBRARY_DIRS:INTERNAL=
PKG_GSTREAMER_STATIC_LIBS:INTERNAL=
PKG_GSTREAMER_STATIC_LIBS_L:INTERNAL=
PKG_GSTREAMER_STATIC_LIBS_OTHER:INTERNAL=
PKG_GSTREAMER_STATIC_LIBS_PATHS:INTERNAL=
PKG_GSTREAMER_VERSION:INTERNAL=1.8.3
PKG_GSTREAMER_gstreamer-1.0_INCLUDEDIR:INTERNAL=
PKG_GSTREAMER_gstreamer-1.0_LIBDIR:INTERNAL=
PKG_GSTREAMER_gstreamer-1.0_PREFIX:INTERNAL=
PKG_GSTREAMER_gstreamer-1.0_VERSION:INTERNAL=
//ADVANCED property for variable: YARP_USE_GSTREAMER
YARP_USE_GSTREAMER-ADVANCED:INTERNAL=1
//ADVANCED property for variable: YARP_USE_GSTREAMERPLUGINSBASE
YARP_USE_GSTREAMERPLUGINSBASE-ADVANCED:INTERNAL=1
//Use system-installed GStreamer, rather than a private copy (recommended)
YARP_USE_SYSTEM_GSTREAMER:INTERNAL=ON
//Use system-installed GStreamerPluginsBase, rather than a private
YARP_USE_SYSTEM_GSTREAMERPLUGINSBASE:INTERNAL=ON
__pkg_config_checked_PKG_GSTREAMER:INTERNAL=1
__pkg_config_checked_PKG_GSTREAMER_APP:INTERNAL=1
__pkg_config_checked_PKG_GSTREAMER_PLUGINS_BASE:INTERNAL=1
```
In the attached file you can find the output of compilation.
[h264CarrierCompilationOutput.txt](https://github.com/robotology/yarp/files/1378498/h264CarrierCompilationOutput.txt)

**Comment by drdanz** [comment]
Hmm I see... the command contains `-isystem /usr/include/gstreamer-1.0` but the file is `/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include/gst/gstconfig.h`, so the `x86_64-linux-gnu` folder is missing...
What ubuntu/debian version are you using?

**Comment by drdanz** [comment]
@valegagge can you also post the output of `dpkg -l libgstreamer1.0-dev` and `dpkg -L libgstreamer1.0-dev`?

**Comment by drdanz** [comment]
I rebased again and (blindly) tried to fix your issue. @valegagge can you try it and tell me if it works?

**Comment by valegagge** [comment]
@drdanz Ok now it works!

**Comment by valegagge** [comment]
Hi @drdanz ,
after your commit h264 carrier works properly. 
@Nicogene  and I cleaned the history and pushed on branch.

**Comment by drdanz** [comment]
Merged, thanks!

**Pull request merged**